include/circt/Dialect/Moore/MooreTypes.h

```diff
@@ -170,31 +147,9 @@ class PackedType : public UnpackedType {
 };
 
 //===----------------------------------------------------------------------===//
-// Packed and Unpacked Structs
+// Struct Members
 //===----------------------------------------------------------------------===//
 
-/// Whether a struct is a `struct`, `union`, or `union tagged`.
-enum class StructKind {
-  /// A `struct`.
-  Struct,
-  /// A `union`.
-  Union,
-  /// A `union tagged`.
-  TaggedUnion,
```

Comment:
After some tests, I found union-tag check will be done in Slang such as 
union.v:46:15: error: value of type 'unionTemp' cannot be assigned to type 'bit[3:0]'
msg1.halfMsg  = myUnion;
error message will be printed.
So pack/unpacked struct/union will be all right.

Reply:
Also in mooreops.td definining ops of procedure also has these kinds
```
def ProcedureKindAttr: I32EnumAttr<"ProcedureKind", "Procedure kind",
            [Initial, Final, Always, AlwaysComb, AlwaysLatch, AlwaysFF]>{
  let cppNamespace = "circt::moore";
}
```
Will it be better to create more ops like procedure_initial, procedure_final and so on ?
Some of them will be lowered to hw but procedure initial will be lowered to llhd.

Reply:
Personally, I suggest let proceudre as a abstract class and let Initial, Final, Always, AlwaysComb, AlwaysLatch, AlwaysFF procedure ops implements separately with def declaration.

Reply:
This is a bit of a balancing act. For types it's often easier to split them apart into separate types, as long as there is no explosion of combinations that you would have to support. This helps with type checks, where you often would prefer to write `isa<UnionType>()` instead of `auto ty = dyn_cast<StructLikeType(); ty && ty.getKind() == StructKind::Union`.

Procedures have very similar structure and limitations. That suggests a single op with an enum kind. You are totally right though, we may want to have more specialized process ops and lower to them. For example, for simple `always_comb` which are the only driver to the assigned signals, we could create something like this:
```
moore.procedure always_comb {
  %0 = comb.add %a, %b
  moore.blocking_assign %x, %0
  %1 = comb.sub %0, %c
  moore.blocking_assign %y, %1
}
// ----- 8< ----- lowered into ----- 8< ----- 
%2, %3 = moore.execute_region {
  %0 = comb.add %a, %b
  %1 = comb.sub %0, %c
  moore.yield %0, %1
}
moore.assign %x, %2
moore.assign %y, %3
```

The idea is to start close to the AST, and then massage the IR to get closer to our goal (inlined always blocks, registers detected and made explicit, etc.) to make MooreToCore's life very easy :smiley: